### ISSUE TITLE
Double precision of digits smaller by half for java bindings vs. python bindings

### ISSUE THREAD

**Issue by geoHeil** [uber/h3-java#40]
```
@transient lazy val h3 = H3Core.newInstance()
val h3INdex = h3.geoToH3(48.209046, 16.368850, 7)
h3INdex: Long = 608515207542603775
scala> h3.h3ToGeoBoundary(h3INdex)
res11: java.util.List[com.uber.h3core.util.GeoCoord] = [GeoCoord{lat=48,218097, lng=16,345586}
```

The number of  digits `48,218097` different from the python-binding there, it has 14 digits: `48.21809738156692` 

How can I get this precision also for the java bindings?

Note I ran these snippets in a scala shell but it should not really make a difference. 

# edit
python code
`h3.h3_to_geo_boundary(h3.geo_to_h3(48.209046, 16.368850, 7))
`outputs:
```
[[48.21809738156692, 16.345586457608764],
 [48.20585865184859, 16.345938048970673],
 [48.20032251826552, 16.362946520461485],
 [48.20702432725267, 16.37960717187623],
 [48.219263502809156, 16.379260992862143],
 [48.224800424038904, 16.362248749695652]]
```

**Comment by isaacbrodsky** [comment]
Hi,

In both cases we should be returning the same double precision floating point numbers. I think the output you're seeing may have to do with how Java prints floats by default. I tried the following code:

```
        long index = h3.geoToH3(48.209046, 16.368850, 7);
        System.out.println(h3.h3ToString(index) + " (in decimal " + index + ")");
        List<GeoCoord> bounds = h3.h3ToGeoBoundary(index);
        System.out.println(bounds);
        System.out.println(new DecimalFormat("0.###############").format(bounds.get(0).lat));
```

which gave me output:

```
871e15b71ffffff (in decimal 608515207542603775)
[GeoCoord{lat=48.218097, lng=16.345586}, GeoCoord{lat=48.205859, lng=16.345938}, GeoCoord{lat=48.200323, lng=16.362947}, GeoCoord{lat=48.207024, lng=16.379607}, GeoCoord{lat=48.219264, lng=16.379261}, GeoCoord{lat=48.224800, lng=16.362249}]
48.21809738156692
```

This seems to match the expected output from Python, the only change being using a decimal formatter with a specific number of digits.

**Comment by geoHeil** [comment]
I need to look at your answer in detail. However, the reason I Initially looked at it was:

- I want to get a WKT LINESTRING for the boundary of a hexagon
- the code below computes such a boundary, but fails with `Points of LinearRing do not form a closed linestring`
- debugging the code outlined that only minimal precision (unlike in python) was being used)
```
def wktBoundary(h3IndexColumn: Long): String = {
    @transient lazy val factory = new GeometryFactory()
    @transient lazy val writer = new WKTWriter()

    import scala.collection.JavaConverters._
    val coordinates = h3.h3ToGeoBoundary(h3IndexColumn).asScala.map(e=> new Coordinate(e.lat, e.lng)).toArray
    coordinates.foreach(println) // small precision
    val polygon = factory.createPolygon(coordinates) // fails to create polygon, linestring not closed.
    writer.write(polygon)
  }
```

in python:
```
from shapely.geometry.polygon import Polygon
Polygon(h3.h3_to_geo_boundary(h3_hex_string))
```
works just fine (and the precision is matching.

So basically I do not understand why in the java / Scala world the Geometry Constructor fails to create the polygon.

# looking at your answer
@isaacbrodsky thanks for detailing this. Internally 
```
public class GeoCoord {
    public final double lat;
    public final double lng;

```
stores Doubles but maybe `import scala.collection.JavaConverters._` has similar format issues with it?

**Comment by dfellis** [comment]
@geoHeil copy the first pair of coordinates and append it to the end of the list. It wants a closed loop and doesn't know how to do that on its own.

**Comment by geoHeil** [comment]
@dfellis ok so manual closing is required for JTS. Thanks. Anyhow applying your suggestion like:

```
def wktBoundary(h3IndexColumn: Long): String = {
    @transient lazy val factory = new GeometryFactory()
    @transient lazy val writer = new WKTWriter()

    import scala.collection.JavaConverters._
//    h3.h3ToGeoBoundary(h3IndexColumn).asScala.map(e=> (e.lat, e.lng)).foreach(println)
    // swap lng, lat here to reflect java geometry construct of x,y (x_lng, y_lat)
    val coordinates = h3.h3ToGeoBoundary(h3IndexColumn).asScala.map(e=> new Coordinate(e.lng, e.lat)).toArray
    val closedRing = coordinates ++ Array(coordinates.head)
    closedRing.foreach(println)
    val polygon = factory.createPolygon(coordinates)
    writer.write(polygon)
  }
  wktBoundary(h3.geoToH3(48.209046, 16.368850, 7))
scala>   wktBoundary(h3.geoToH3(48.209046, 16.368850, 7))
(16.345586457608764, 48.21809738156692, NaN)
(16.345938048970673, 48.20585865184859, NaN)
(16.362946520461485, 48.20032251826552, NaN)
(16.37960717187623, 48.20702432725267, NaN)
(16.379260992862143, 48.219263502809156, NaN)
(16.362248749695652, 48.224800424038904, NaN)
(16.345586457608764, 48.21809738156692, NaN)
java.lang.IllegalArgumentException: Points of LinearRing do not form a closed linestring

```
duplicates the first coordinate pair (and appends it at the end. It should close the geometry (start and end point the same).
Still the ring is not closed.

**Comment by dfellis** [comment]
You've changed a couple things at the same time here.

1. Instead of a simple array of arrays, it's an array of coordinate objects, which apparently also want an altitude parameter that is currently NaN -- perhaps that needs to be explicitly set to zero?
2. You've swapped the order from (lat, lng) to (lng, lat) which order does the `GeometryFactory` want them?

I don't think there is anything that the H3 library is doing wrong here, as Isaac demonstrated. The WKT documentation will have the answer to what you're looking for, but duplicating the first coordinates at the end of the list *does* close the polygon. I suspect the `NaN` in those coordinates is what's causing trouble now.
